Refactor deserialization to be more compact and a single function

Geod24 opened the pull request:
Also add a few features, such as full support for range, support for bool, validation for strings... See each commits.

Geod24 committed:
Add ability to [de]serialize a range with length

Simplify (de)serialization of types by providing a generic way
to deal with ranges with length (mainly arrays, but other range qualify).

Geod24 committed:
Change deserializeVarInt to return the value directly

Also removes a redundant assert, asunsigned types are always >= 0,
and modify the final branch to be more obvious.
Ideally the compiler should see that the whole ubyte range is covered,
so the assert(0) shouldn't be necessary, but it currently does not.

Geod24 committed:
Deserializer: Merge scalar types handling, support 'bool'

Geod24 committed:
Merge all templated deserialization routines into one

It is becoming increasingly more difficult to make deserialization
work in all possible cases, as many types can match multiple template
definitions, leading to an error.

While overload sets are a good way to deal with this,
they also require all supported types to be explicitly listed.

Merging those functions in one will hopefully make code clearer
and the priority easier to reason about.

Geod24 committed:
Deserializer: Integrate array deserialization

Make sure that deserialized arrays are processed in batch,
not element by element.
Additionally, validate UTF data whenever a string is received.

AndrejMitrovic commented on source/agora/common/Deserializer.d:
```diff
@@ -84,6 +84,10 @@ unittest
 public alias DeserializeDg = ubyte[] delegate(size_t size) @safe;
 
+/// Traits to check if a given type has an in place deserialization routine
+private enum hasDeserializeMethod (T) = is(T == struct)
```
@bpfkorea/core in case you are confused about this weird looking declaration, this is just a shorter way of writing a full template declaration like this:

```D
template hasDeserializeMethod(T)
{
    enum hasDeserializeMethod = is(T == struct) &&
        is(typeof(T.init.deserialize(DeserializeDg.init)));
}
```

See https://dlang.org/spec/template.html#implicit_template_properties for more info.

AndrejMitrovic commented on source/agora/common/Deserializer.d:
```diff
@@ -280,18 +275,24 @@ public void deserializePart (ref Hash[] record, scope DeserializeDg dg)
+
+    Throws:
+      If the deserialized value does not fit into a `T`.
+      Not that for `ulong`, this function is `nothrow`.
```
```suggestion
      Note that for `ulong`, this function is `nothrow`.
```

AndrejMitrovic commented on source/agora/common/Deserializer.d:
```diff
@@ -168,48 +169,30 @@ public void deserializePart (T)(ref T record, scope DeserializeDg dg)
-{
-    if (compact == CompactMode.Yes)
-        record = deserializeVarInt!ulong(dg);
+    static if (is(Unqual!T == bool))
```
Was `bool` not supported before? Might be good to add a test.

Geod24 replied:
Not it wasn't because it needs `!!` (`ubyte` does not convert to `bool` implicitly). Will add test.

AndrejMitrovic commented on source/agora/common/Serializer.d:
```diff
@@ -67,18 +68,57 @@ private enum hasSerializeMethod (T) = is(T == struct)
 
 public ubyte[] serializeFull (T) (scope const auto ref T record)
     @safe
-    if (is(T == struct))
```
Will this end up enabling serialization for classes too?

Geod24 replied:
Hum yes although that wasn't the point.

Geod24 replied:
[Aaaaaaaaaacccckkkkktually...](https://github.com/bpfkorea/agora/blob/c0b4e7ba5e72dafab168238439e09df3b5b3e37d/source/agora/common/BanManager.d#L266-L311)

AndrejMitrovic commented on source/agora/common/Deserializer.d:
```diff
@@ -149,61 +149,63 @@ public T deserializeFull (T) (scope DeserializeDg dg) @safe
+    {
+        OriginalType!T orig_val;
+        deserializePart(orig_val, dg);
+        record = cast(T)(orig_val);
```
What about `assert(orig_val >= T.min && orig_val <= T.max);`? Although I'm not sure if this works well with things like `enum E { f = "foo", b = "bar" }`..

Geod24 replied:
I didn't wanna make any assumption about the type of `enum`. I don't think we support enum strings though, only enum based on integer. There are quite a few places where we need to sanitize the input: most notably, length of array (someone could DoS us with just a bit of binary data...). Maybe we should open an issue for this ?

AndrejMitrovic commented on source/agora/common/Deserializer.d:
```diff
@@ -286,7 +288,9 @@ private T deserializeVarInt (T) (scope DeserializeDg dg)
     }
 
-    if (int_size <= 0xFC)
+    static if (is(Unqual!T == ubyte))
```
Did you add this to evade a cast? But wouldn't the compiler already know that a cast from ubyte to ubyte is just a noop? Curious.

Geod24 replied:
No, i actually wrote it because `ubyte` are not size-encoded (it doesn't make sense). But to keep the code generic, this function is always called. Come to think of it, modifying the call site might be better...

Geod24 commented:
Updated:
- Less commits
- Added a test for `bool`
- Reverted the `ubyte` change to `deserializeVarInt`
- Fixed `Note` typo;

Comments left "unaddressed":
- https://github.com/bpfkorea/agora/pull/539#discussion_r379246181 (was a question)
- https://github.com/bpfkorea/agora/pull/539#discussion_r379246996 (enums are weird and not well supported so far)

AndrejMitrovic commented:
@bpfkorea/core have you looked at this, does it make sense to you?

AndrejMitrovic commented:
I guess nobody disagrees with the changes. Merging..
